### PR TITLE
Module experiments: Add one more prelude layer for extern crate names passed with `--extern`

### PR DESCRIPTION
**Pull request by petrochenkov**
Implements one item from https://internals.rust-lang.org/t/the-great-module-adventure-continues/6678/183

When some name is looked up in lexical scope (`name`, i.e. not module-relative scope `some_mod::name` or `::name`), it's searched roughly in the next order:
- local variables
- items in unnamed blocks
- items in the current module
- :sparkles: NEW! :sparkles: crate names passed with `--extern` ("extern prelude")
- standard library prelude (`Vec`, `drop`)
- language prelude (built-in types like `u8`, `str`, etc)

The last two layers contain a limited set of names controlled by us and not arbitrary user-defined names like upper layers. We want to be able to add new names into these two layers without breaking user code, so "extern prelude" names have higher priority than std prelude and built-in types.
This is a one-time breaking change, that's why it would be nice to run this through crater.
Practical impact is expected to be minimal though due to stylistic reasons (there are not many `Uppercase` crates) and due to the way how primitive types are resolved (https://github.com/rust-lang/rust/pull/32131).

### CODE REVIEW

**Comment by petrochenkov** [comment]
Needs a check-only crater run.

**Comment by nikomatsakis** [comment]
> We want to be able to add new names into these two layers without breaking user code, so "extern prelude" names have higher priority than std prelude and built-in types.

Ah, I was going to ask why you did it that way -- but of course, that makes sense.

**Comment by nikomatsakis** [comment]
r=me pending crater results

**Comment by cramertj** [comment]
@Mark-Simulacrum Are the crater results for this available yet?

**Comment by Mark-Simulacrum** [comment]
Unfortunately, they aren't; I'd anticipate about 4 more days. This wasn't marked as check-only in the crater spreadsheet so it was started as a normal run :/

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -1972,13 +1974,31 @@ impl<'a> Resolver<'a> {
...
-                    .ok().map(LexicalScopeBinding::Item)
+        if !module.no_implicit_prelude {
+            if ns == TypeNS && self.extern_prelude.contains(&ident.name) {
+                if !self.session.features_untracked().extern_prelude {
+                    feature_err(&self.session.parse_sess, "extern_prelude",
+                                ident.span, GateIssue::Language,
+                                "access to extern crates through prelude is experimental").emit();
+                }
+
+                let crate_id = self.crate_loader.resolve_crate_from_path(ident.name, ident.span);
```
Comment (Manishearth) [inline src/librustc_resolve/lib.rs]: Can we use process_path_extern here?

File: src/test/run-make-fulldeps/extern-prelude/Makefile [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+all:
+	$(RUSTC) ep-lib.rs
+	$(RUSTC) ep-vec.rs
+
+	$(RUSTC) basic.rs --extern ep_lib=$(TMPDIR)/libep_lib.rlib
+	$(RUSTC) shadow-mod.rs --extern ep_lib=$(TMPDIR)/libep_lib.rlib
+	$(RUSTC) shadow-prelude.rs --extern Vec=$(TMPDIR)/libep_vec.rlib
+	$(RUSTC) feature-gate.rs --extern ep_lib=$(TMPDIR)/libep_lib.rlib 2>&1 | $(CGREP) "access to extern crates through prelude is experimental"
+	$(RUSTC) relative-only.rs --extern ep_lib=$(TMPDIR)/libep_lib.rlib 2>&1 | $(CGREP) "unresolved import"
+	$(RUSTC) relative-only.rs --extern ep_lib=$(TMPDIR)/libep_lib.rlib 2>&1 | $(CGREP) "failed to resolve"
```
Comment (nikomatsakis) [inline src/test/run-make-fulldeps/extern-prelude/Makefile]: why is this using `run-make` and not auxiliary? ah, well, I guess we'd have to modify the `auxiliary` mechanism to supply `--extern`? (doesn't seem that hard... but maybe we leave it for another PR)

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@Mark-Simulacrum ping re: crater

**Comment by Mark-Simulacrum** [comment]
Crater is done, sorry for the delay: http://cargobomb-reports.s3.amazonaws.com/pr-49789/index.html.

**Comment by petrochenkov** [comment]
There are a lot of regression, but all regressions I've checked are caused by something that looks like an implementation bug - `iovec` in `libc::iovec` is interpreted as a crate `iovec`.
Why on earth `resolve_ident_in_lexical_scope` is even called for something that is not in lexical scope?
I need to investigate.

**Comment by Manishearth** [comment]
I think the issue is that the `unused_qualifications` lint is barging in and trying to resolve this against the root, hitting the feature gate. We should probably silently return None when stuff is being resolved speculatively.

**Comment by Manishearth** [comment]
Actually, can we just _not_ throw feature errors in this case? Ideally the module system would pervasively track when it's resolving user-defined paths vs speculative resolution (in case we wish to lint or throw better messages). But this ideally will be done by creating a better abstraction for paths over `&[Ident]`. Due to https://github.com/rust-lang/rust/pull/50100 we already track this in part with the `node_id` argument to `resolve_path` but that isn't perfect.

I'm going to remove the feature error here ; we have plenty of speculative resolution happening in resolve. A longer term solution would be to implement said abstraction (something that contains an `&[Ident]` and `Option<NodeId>`) -- this can be mentored.

**Comment by Manishearth** [comment]
Rebased and pushed an update.

**Comment by Manishearth** [comment]
Should we cargobomb again?

(Sorry for pushing over your PR, @petrochenkov, I was asked to look into this during the edition meeting and we're trying to get edition stuff landed quickly so I just pushed)

**Comment by Manishearth** [comment]
r? @nikomatsakis 

I'll push up a test soonish.

**Comment by Manishearth** [comment]
Removing the feature gate error breaks the test that tests for the feature
gate error, sounds about right.

**Comment by Manishearth** [comment]
Added test, removed gating test.

**Comment by nikomatsakis** [comment]
r=me but I'd like @petrochenkov to sign off on it I think

**Comment by petrochenkov** [comment]
Hmm, this needs another check-only crater run.
With so many regressions caused by the speculative resolution bug we don't know the number of actual regressions.

**Comment by Manishearth** [comment]
@Mark-Simulacrum @kennytm crater ping?

We were thinking we'd land first since it's now structured so that it shouldn't affect non-edition code, and we want to get edition stuff landed quickly. Once landed we can also mess with crater with the edition flag flipped to see the actual breakage.

**Comment by Mark-Simulacrum** [comment]
Another Crater run started, should be done 3 days from now.

**Comment by nikomatsakis** [comment]
I just want to be clear on what is being tested:

Are we testing the effect of adding the names to the prelude? (Presumably yes?)

**Comment by Manishearth** [comment]
We're testing that this doesn't break anything, when off. I think we'll be doing another crater run after this lands with the edition flag flipped? Idk.

**Comment by petrochenkov** [comment]
@Manishearth, what are you doing?
I pushed a proper fix for the speculative resolution issue 3 days ago, but now it's reverted and the PR is reset to your branch.
At least the crater run started before that.

**Comment by petrochenkov** [comment]
I've re-pushed my version, now this PR matches what is tested by crater.

**Comment by Manishearth** [comment]
Ah looks like you pushed after I'd left my comment but before I'd pushed, and my push overwrote it. Sorry about that.

**Comment by nikomatsakis** [comment]
@petrochenkov ok, I see -- so point is, we are testing with the feature *enabled*, right? Sounds good!

**Comment by Mark-Simulacrum** [comment]
Crater run done - http://cargobomb-reports.s3.amazonaws.com/pr-49789-2/index.html. All regressions are spurious.

**Comment by petrochenkov** [comment]
Excellent!
@bors r=nikomatsakis

**Comment by petrochenkov** [comment]
Looks like a spurious network failure
@bors retry

**Comment by aturon** [comment]
Woohoo! Thanks @petrochenkov and @Manishearth for seeing this through!

**Comment by nikomatsakis** [comment]
Is this in the latest nightly? If so, it doesn't seem to be working for me (this commit doesn't build):

https://github.com/nikomatsakis/borrow-check/commit/fd3e48fdf49ac75091f3d8f3cda677dc5a30ee17

I see:

```
-*- mode: compilation; default-directory: "~/versioned/borrow-check/" -*-
Compilation started at Thu May  3 06:43:32

cargo test
   Compiling assert_cli v0.5.4
   Compiling borrow-check v0.1.0 (file:///home/nmatsakis/versioned/borrow-check)
error[E0432]: unresolved import `abomonation_derive`
 --> src/facts.rs:1:5
  |
1 | use abomonation_derive::Abomonation;
  |     ^^^^^^^^^^^^^^^^^^ Maybe a missing `extern crate abomonation_derive;`?

error[E0432]: unresolved import `fxhash`
  --> src/output/mod.rs:14:5
   |
14 | use fxhash::FxHashMap;
   |     ^^^^^^ Maybe a missing `extern crate fxhash;`?

error[E0432]: unresolved import `fxhash`
 --> src/output/dump.rs:3:5
  |
3 | use fxhash::FxHashMap;
  |     ^^^^^^ Maybe a missing `extern crate fxhash;`?

error[E0433]: failed to resolve. Maybe a missing `extern crate differential_dataflow;`?
  --> src/output/timely.rs:15:5
   |
15 | use differential_dataflow::collection::Collection;
   |     ^^^^^^^^^^^^^^^^^^^^^ Maybe a missing `extern crate differential_dataflow;`?

...
```

**Comment by petrochenkov** [comment]
@nikomatsakis 
Prelude doesn't affect resolution for absolute paths (including imports).
Resolve-the-first-segment-as-extern-crate for absolute paths is enabled by `feature(extern_absolute_paths)`.

**Comment by nikomatsakis** [comment]
Hmm, it seems that the commits appear in the git log, so maybe I'm doing something wrong?

**Comment by nikomatsakis** [comment]
Ah, I see, duh. Thanks!
